### PR TITLE
update OTLP proto to v0.9.0

### PR DESCRIPTION
**Pull request by codeboten**
**Description:**
Update opentelemetry-proto to v0.9.0. As part of the change:
- bumped submodule
- updated sed patch file to support deprecated nullable `StringKeyValue`
- added translation for labels to attributes in otlp_wrapper

**Link to tracking Issue:** Part of https://github.com/open-telemetry/opentelemetry-collector/issues/3535

This is built on top of https://github.com/open-telemetry/opentelemetry-collector/pull/3735

**Remaining work once this PR is done:**
- [ ] convert labels to attributes
- [ ] schema_url
- [ ] add bytes data type

### CODE REVIEW

**Review by tigrannajaryan — COMMENTED** [review]
Looking at the changeset https://github.com/open-telemetry/opentelemetry-proto/compare/v0.8.0...v0.9.0 it seems like there are a couple more things that will need to be added:
- schema_url
- bytes data type
Will these be added in a future PR? (I think that would be best to keep this one small).

**Comment by codeboten** [comment]
> Looking at the changeset [open-telemetry/opentelemetry-proto@v0.8.0...v0.9.0](https://github.com/open-telemetry/opentelemetry-proto/compare/v0.8.0...v0.9.0) it seems like there are a couple more things that will need to be added:
> 
>     * schema_url
> 
>     * bytes data type
>       Will these be added in a future PR? (I think that would be best to keep this one small).

Yes that's correct, I'm only trying to do the bare minimum to get the tests passing w/ OTLP v0.9.0, I've added the rest to a task list on this PR's description, though they should probably go in the main issue https://github.com/open-telemetry/opentelemetry-collector/issues/3535, but I don't have permission to add tasks there.

File: model/pdata/metrics.go [hunk 1]
```diff
@@ -129,9 +125,7 @@ type MetricDataType int32
 
 const (
 	MetricDataTypeNone MetricDataType = iota
-	MetricDataTypeIntGauge
```
[the remark this answers is not in the document]
Comment (codeboten) [reply model/pdata/metrics.go:132]: Yes these have been removed as part of the upgrade work to OTLP v0.8.0

**Comment by bogdandrutu** [comment]
Please rebase, and make more granular changes :)

**Comment by codeboten** [comment]
> Please rebase, and make more granular changes :)

@bogdandrutu done, will add separate PRs for converting labels to attributes.

File: proto_patch.sed [hunk 2]
```diff
@@ -27,9 +27,11 @@ s+repeated KeyValue \(.*\);+repeated KeyValue \1\
 s+AnyValue \(.*\);+AnyValue \1\
   [ (gogoproto.nullable) = false ];+g
 
-s+repeated opentelemetry.proto.common.v1.StringKeyValue \(.*\);+repeated opentelemetry.proto.common.v1.StringKeyValue \1\
+s+repeated opentelemetry.proto.common.v1.StringKeyValue \(.*\)\(?<!\]\);+repeated opentelemetry.proto.common.v1.StringKeyValue \1\
   [ (gogoproto.nullable) = false ];+g
 
+s+repeated opentelemetry.proto.common.v1.StringKeyValue \(.*\)\];+repeated opentelemetry.proto.common.v1.StringKeyValue \1\, (gogoproto.nullable) = false ];+g
+
```
[the remark this answers is not in the document]
Comment (codeboten) [reply proto_patch.sed]: Added a comment.

File: model/internal/otlp_wrapper.go [hunk 3]
```diff
@@ -44,6 +45,7 @@ func MetricsFromOtlp(req *otlpcollectormetrics.ExportMetricsServiceRequest) Metr
 // - Convert IntHistogram to Histogram. See https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L170
 // - Convert IntGauge to Gauge. See https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L156
 // - Convert IntSum to Sum. See https://github.com/open-telemetry/opentelemetry-proto/blob/f3b0ee0861d304f8f3126686ba9b01c106069cb0/opentelemetry/proto/metrics/v1/metrics.proto#L156
+// - Converts Labels to Attributes. See https://github.com/open-telemetry/opentelemetry-proto/blob/8672494217bfc858e2a82a4e8c623d4a5530473a/opentelemetry/proto/metrics/v1/metrics.proto#L385
```
Comment (bogdandrutu) [inline model/internal/otlp_wrapper.go:48]: Can we do this in a separate PR?

Comment (codeboten) [reply model/internal/otlp_wrapper.go:48]: Ah yes, this comment was leftover from reverting changes earlier.

File: model/internal/otlp_wrapper.go [hunk 4]
```diff
@@ -125,7 +126,7 @@ func intHistogramToHistogram(src *otlpmetrics.Metric_IntHistogram) *otlpmetrics.
 	datapoints := []*otlpmetrics.HistogramDataPoint{}
 	for _, datapoint := range src.IntHistogram.DataPoints {
 		datapoints = append(datapoints, &otlpmetrics.HistogramDataPoint{
-			Labels:            datapoint.Labels,
+			Labels:            stringKeyValuePtrArrayToStringKeyValueArray(datapoint.Labels),
```
Comment (tigrannajaryan) [inline model/internal/otlp_wrapper.go:129]: Sorry, I don't understand why does the need to do this arise? Why aren't all StringKeyValue pair lists using either slices of pointers or slices of values? Why do we need to mix and then be forced to do this conversion?

File: model/internal/data/protogen/metrics/v1/metrics.pb.go [hunk 5]
```diff
@@ -893,21 +940,14 @@ func (m *Summary) GetDataPoints() []*SummaryDataPoint {
...
 type IntDataPoint struct {
 	// The set of labels that uniquely identify this timeseries.
-	Labels []v11.StringKeyValue `protobuf:"bytes,1,rep,name=labels,proto3" json:"labels"`
-	// start_time_unix_nano is the last time when the aggregation value was reset
-	// to "zero". For some metric types this is ignored, see data types for more
-	// details.
-	//
-	// The aggregation value is over the time interval (start_time_unix_nano,
-	// time_unix_nano].
+	Labels []*v11.StringKeyValue `protobuf:"bytes,1,rep,name=labels,proto3" json:"labels,omitempty"`
```
Comment (tigrannajaryan) [inline model/internal/data/protogen/metrics/v1/metrics.pb.go:943]: Why did this need to change? Can we keep `Labels []v11.StringKeyValue` and avoid the need to have `stringKeyValuePtrArrayToStringKeyValueArray` func?

Comment (codeboten) [reply model/internal/data/protogen/metrics/v1/metrics.pb.go:943]: 🤔 that's a good question, this is all generated via `make genproto`, will take a look at why this happened.

Comment (codeboten) [reply model/internal/data/protogen/metrics/v1/metrics.pb.go:943]: Thanks for catching this, it was a problem with the sed patch not replacing `StringKeyValue` correctly, fixing now.

File: model/internal/otlp_wrapper.go (hunk 4 above)
Comment (codeboten) [reply model/internal/otlp_wrapper.go]: This is now fixed.

File: model/internal/otlp_wrapper.go [hunk 6]
```diff
@@ -196,3 +236,30 @@ func intExemplarToExemplar(src []otlpmetrics.IntExemplar) []otlpmetrics.Exemplar
 	}
 	return exemplars
 }
+
+func numberDataPointsLabelsToAttributes(dps []*otlpmetrics.NumberDataPoint) {
+	for i, _ := range dps {
+		if dps[i].Labels != nil {
+			dps[i].Attributes = labelsToAttributes(dps[i].Labels)
```
[the remark this answers is not in the document]
Comment (codeboten) [reply model/internal/otlp_wrapper.go]: This has been moved out of this PR, but I think it's an important point to clarify. Based on the comments in the proto, https://github.com/open-telemetry/opentelemetry-proto/blob/48ca003a72365f5591f1acd9b065ea8316be2ade/opentelemetry/proto/metrics/v1/metrics.proto#L307-L312 what are your thoughts on the right way to move forward? I suppose we can either:

1. assume the Attributes are empty and override them with labels
2. upsert Labels into Attributes
3. insert Labels that are not present only

3 seems to make the most sense to me, but it's not a strong preference.

**Review by bogdandrutu — APPROVED** [review]

File: model/internal/otlp_wrapper.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (bogdandrutu) [reply model/internal/otlp_wrapper.go]: I would say we need to assume Attributes is empty:
Users cannot constructor custom protos since they are hidden. This case can happen only when we receive from the wire. So `MetricsCompatibilityChanges` is only called from data received on the wire which can be in new or old format but not a mix.

Depends how backwards compatible we want to be:
1. If only labels is set we upsert everything to the Attributes
2. If both are received the expectation is that they are the same, do nothing
3. If only attributes is received then do nothing.

On the sent side:
* If we want to do the correct thing we should translate attributes to labels and send both, but this is not what we did with IntSum/IntGauge. So maybe just a CHANGELOG entry and say that we broke this?

[the remark this answers is not in the document]
Comment (codeboten) [reply model/internal/otlp_wrapper.go]: Got it.

When receiving:
- we should check if both labels and attributes are set, if so, don't do anything
- if labels are set, upsert into attributes and leave labels as is. 
- don't do anything with the attributes only case.

When exporting, should labels always be dropped here? I worry that we would end up with inconsistent behaviour if labels are sometimes set and sometimes not.

File: model/internal/data/protogen/metrics/v1/metrics.pb.go [hunk 7]
```diff
@@ -990,22 +1030,25 @@ func (m *IntDataPoint) GetExemplars() []IntExemplar {
...
+	// 1. Old senders and receivers that are not aware of this change will
+	// continue using the `labels` field.
+	// 2. New senders, which are aware of this change MUST send only `attributes`.
+	// 3. New receivers, which are aware of this change MUST convert this into
+	// `labels` by simply converting all int64 values into float.
 	//
-	// The aggregation value is over the time interval (start_time_unix_nano,
-	// time_unix_nano].
+	// This field will be removed in ~3 months, on July 1, 2021.
+	Labels []v11.StringKeyValue `protobuf:"bytes,1,rep,name=labels,proto3" json:"labels"` // Deprecated: Do not use.
```
Comment (CoderPoet) [inline model/internal/data/protogen/metrics/v1/metrics.pb.go:1044]: Has the Labels field been removed in the latest version? We still use the old version of the collector in some places. Is there any way to make it compatible?
